### PR TITLE
React 19 support

### PR DESCRIPTION
**Pull request by snowystinger**
Closes https://github.com/adobe/react-spectrum/issues/6445

Types for React 19 have not been released yet, but can work off the pre-release following https://github.com/eps1lon/react-spectrum/pull/1

## Unresolved issues
framer-motion uses the deprecated, global JSX namespace. Tracking in https://github.com/eps1lon/DefinitelyTyped/issues/25
lucide-react uses removed ReactSVG type. Tracking in https://github.com/eps1lon/DefinitelyTyped/issues/25

Dev docs build: https://github.com/mdx-js/mdx/issues/2487
markdown-to-jsx: https://github.com/quantizor/markdown-to-jsx/issues/583

Need to upgrade to Storybook 8, might address some type issues
Testing-library-react might need an upgrade

We can probably patch-package most of these TS errors if we need
On another note, the whole JSX global namespace removal is causing all kinds of issues apparently https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69142

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,12 +12,12 @@
     "check-types": "tsc && tsc-strict",
     "install-16": "yarn add -W react@^16.8.0 react-dom@^16.8.0 @testing-library/react@^12 @testing-library/react-hooks@^8 @testing-library/dom@^8 && node scripts/oldReactSupport.mjs",
     "install-17": "yarn add -W react@^17 react-dom@^17 @testing-library/react@^12 @testing-library/react-hooks@^8 @testing-library/dom@^8 && node scripts/oldReactSupport.mjs",
-    "install-canary": "yarn add -W react@canary react-dom@canary",
+    "install-19": "node ./scripts/install-react-19-types.mjs && yarn add -W react@next react-dom@next",
```
Comment (snowystinger) [inline package.json:15]: should types for 19 always be installed?

**Comment by snowystinger** [comment]
Rob-bot:

Verdaccio builds:
      [CRA Test App](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/build/index.html)
      [NextJS Test App](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/next/index.html)
      [RAC Tailwind Example](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/rac-tailwind/index.html)
      [RAC Spectrum + Tailwind Example](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/rac-spectrum-tailwind/index.html)
      [CRA Test App Size](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/publish-stats/build-stats.txt)
      [NextJS App Size](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/publish-stats/next-build-stats.txt)
      [Publish stats](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/publish-stats/publish.json)
      [Size diff since last release](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/publish-stats/size-diff.txt)
      [Docs](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/verdaccio/docs/index.html)
      [Storybook](https://reactspectrum.blob.core.windows.net/reactspectrum/3710e8eb71fe90fa9f8b061f7badec7903c98440/storybook/index.html)

File: packages/@react-spectrum/card/test/CardView.test.js [hunk 2]
```diff
@@ -1199,7 +1199,12 @@ describe('CardView', function () {
       let grid = tree.getByRole('grid');
       grid.scrollTop = 3000;
       fireEvent.scroll(grid);
-      expect(onLoadMore).toHaveBeenCalledTimes(1);
+      let isReact19 = parseInt(React.version, 10) >= 19;
+      if (isReact19 && layout !== GridLayout) {
+        expect(onLoadMore).toHaveBeenCalledTimes(2);
```
Comment (devongovett) [inline packages/@react-spectrum/card/test/CardView.test.js:1204]: ?

File: packages/@react-spectrum/link/src/Link.tsx [hunk 3]
```diff
@@ -13,13 +13,14 @@
...
 import {mergeProps, mergeRefs} from '@react-aria/utils';
-import React, {JSX, useRef} from 'react';
+import React, {ForwardedRef, JSX, MutableRefObject, useRef} from 'react';
 import {SpectrumLinkProps} from '@react-types/link';
 import styles from '@adobe/spectrum-css-temp/components/link/vars.css';
 import {useHover} from '@react-aria/interactions';
 import {useLink} from '@react-aria/link';
 import {useProviderProps} from '@react-spectrum/provider';
 
+let isOldReact = React.version.startsWith('16.') || React.version.startsWith('17.') || React.version.startsWith('18.');
```
Comment (devongovett) [inline packages/@react-spectrum/link/src/Link.tsx:23]: `parseInt(React.version) <= 18`?

File: packages/@react-spectrum/list/test/ListViewDnd.test.js [hunk 4]
```diff
@@ -428,10 +429,15 @@ describe('ListView', function () {
         fireEvent(grid, new DragEvent('drop', {dataTransfer, clientX: 1, clientY: 110}));
         act(() => jest.runAllTimers());
         await act(async () => Promise.resolve());
+
         expect(onDrop).toHaveBeenCalledTimes(1);
 
         fireEvent(cell, new DragEvent('dragend', {dataTransfer, clientX: 1, clientY: 110}));
-        expect(onDragEnd).toHaveBeenCalledTimes(1);
+        if (isReact19) {
+          expect(onDragEnd).toHaveBeenCalledTimes(2);
```
Comment (devongovett) [inline packages/@react-spectrum/list/test/ListViewDnd.test.js:437]: I guess this is the react bug we filed? Should we have a todo somewhere to fix it later?

File: packages/@react-spectrum/card/test/CardView.test.js (hunk 2 above)
Comment (snowystinger) [reply packages/@react-spectrum/card/test/CardView.test.js:1204]: these loadmore checks have always been sensitive to any change, I've checked it in the browser and didn't notice an issue
I'll put in a todo to address it, though almost anything with an isOldReact or isReact19 check should be updated/looked into more at some point

**Review by devongovett — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]

**Pull request merged**